# Assignment 6

# Python Basic - 1

**Q.1. What are keywords in python? Using the keyword library, print all the python keywords.**

**Answer:** Keywords are the reserved words in Python. We cannot use a keyword as a variable name, function name or any other identifier.
**1.Value Keywords:** True, False, None.
**2.Operator Keywords:** and, or, not, in, is.
**3.Control Flow Keywords:** if, elif, else.
**4.Iteration Keywords:** for, while, break, continue, else.
**5.Structure Keywords:** def, class, with, as, pass, lambda.
**6.Returning Keywords:** return, yield.
**7.Import Keywords:** import, from, as.

In [1]:
# importing keyword library
import keyword
print("Here's a list of all keywords in Python Programming:...")
# Printing all keywords.
print(keyword.kwlist)

Here's a list of all keywords in Python Programming:...
['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


**Q.2. What are the rules to create variables in python?**

**Answer:**Rules for Python variables:

1.A variable name must start with a letter or the underscore character<br>
2.A variable name cannot start with a number<br>
3.A variable name can only contain alpha-numeric characters and underscores (A-z, 0-9, and _ )<br>
4.Variable names are case-sensitive (age, Age and AGE are three different variables)<br>
5.Do not allow Whitespaces between variable characters. 
6.We cannot use special symbols like !, @, #, $, and so on.

In [ ]:
#Legal variable names:
myvar = "Shiv"
my_var = "Shiv"
_my_var = "Shiv"
myVar = "Shiv"
MYVAR = "Shiv"
myvar2 = "Shiv"

#Illegal variable names:
2myvar = "Priya"
my-var = "Priya"
my var = "Priya"

**Q.3. What are the standards and conventions followed for the nomenclature of variables in
python to improve code readability and maintainability?**

**Answer:** Coding standards are a set of guidelines to produce maintainable and scalable code to ensure code quality, readability, and re-usability.

**Standards and convenstion of variable in python :<br>**
**1.Naming conventions for local variables, global variables, constants and functions:**<br>
Some of the variable naming conventions are given below:<br>
1.Meaningful and understandable variables name helps anyone to understand the reason of using it.<br>
2.Local variables should be named using camel case lettering starting with small letter (e.g. localData) whereas Global variables names should start with a capital letter (e.g. GlobalData). Constant names should be formed using capital letters only (e.g. CONSDATA).<br>
3.It is better to avoid the use of digits in variable names.<br>
4.The names of the function should be written in camel case starting with small letters.<br>
5.The name of the function must describe the reason of using the function clearly and briefly.<br>

**Some of the coding standards are given below:**<br>
**1.Limited use of globals:**<br>
These rules tell about which types of data that can be declared global and the data that can’t be.<br>

**2.Standard headers for different modules:**<br>
For better understanding and maintenance of the code, the header of different modules should follow some standard format and information. The header format must contain below things that is being used in various companies:<br>
1.Name of the module<br>
2.Date of module creation<br>
3.Author of the module<br>
4.Modification history<br>
5.Synopsis of the module about what the module does<br>
6.Different functions supported in the module along with their input output parameters<br>
7.Global variables accessed or modified by the module<br>

**3.Indentation:**<br>
Proper indentation is very important to increase the readability of the code. For making the code readable, programmers should use White spaces properly. Some of the spacing conventions are given below:<br>
1.There must be a space after giving a comma between two function arguments.<br>
2.Each nested block should be properly indented and spaced.<br>
3.Proper Indentation should be there at the beginning and at the end of each block in the program.<br>
4.All braces should start from a new line and the code following the end of braces also start from a new line.<br>

**4.Error return values and exception handling conventions:**<br>
All functions that encountering an error condition should either return a 0 or 1 for simplifying the debugging.<br>

On the other hand, Coding guidelines give some general suggestions regarding the coding style that to be followed for the betterment of understandability and readability of the code. Some of the coding guidelines are given below :<br>

**1.Avoid using a coding style that is too difficult to understand:**<br>
Code should be easily understandable. The complex code makes maintenance and debugging difficult and expensive.<br>


**2.Avoid using an identifier for multiple purposes:**<br>
Each variable should be given a descriptive and meaningful name indicating the reason behind using it. This is not possible if an identifier is used for multiple purposes and thus it can lead to confusion to the reader. Moreover, it leads to more difficulty during future enhancements.<br>


**3.Code should be well documented:**<br>
The code should be properly commented for understanding easily. Comments regarding the statements increase the understandability of the code.<br>


**4.Length of functions should not be very large:**<br>
Lengthy functions are very difficult to understand. That’s why functions should be small enough to carry out small work and lengthy functions should be broken into small ones for completing small tasks.<br>


**5.Try not to use GOTO statement:**<br>
GOTO statement makes the program unstructured, thus it reduces the understandability of the program and also debugging becomes difficult.


**Q.4. What will happen if a keyword is used as a variable name?**

**Answer:**
There are two main reasons:

It makes the developer's life hard; and It makes the compiler's life hard.

If keywords were allowed as variable names, it would be very hard to tell (for the developers and the compilers) whether something was a variable or a keyword. For example, what does the following mean?

if(x == 10)

Is it an if-statement, or calling a function called if?

Neither the developer nor the compiler would be able to tell.


**Q.5. For what purpose def keyword is used?**

**Answer:** Python def keyword is used to define a function, it is placed before a function name that is provided by the user to create a user-defined function.<br>
**There are two types of function in Python programming:**<br>
**Standard library functions** - These are built-in functions in Python that are available to use.<br>
**User-defined functions** - We can create our own functions based on our requirements.

In [ ]:
def function_name(arguments):
    # function body 

    return

Here,<br>
**def **- keyword used to declare a function<br>
**function_name** - any name given to the function<br>
**arguments** - any value passed to function<br>
**return (optional)** - returns value from a function<br>

**Example:**

In [3]:
# function with two arguments
def add_numbers(num1, num2):
    sum = num1 + num2
    print("Sum: ",sum)

# function call with two values
add_numbers(5, 4)

Sum:  9


In [4]:
# function definition
def find_square(num):
    result = num * num
    return result

# function call
square = find_square(3)

print('Square:',square)

Square: 9


**Q.6. What is the operation of this special character ‘\’?**

**Answer:** In Python strings, the backslash " \ " is a special character, also called the "escape" character. It is used in representing **certain whitespace** characters: "\t" is a tab, "\n" is a newline, and "\r" is a carriage return.

**Q.7. Give an example of the following conditions:**<br>
**(i) Homogeneous list**<br>
**(ii) Heterogeneous set**<br>
**(iii) Homogeneous tuple**<br>

**Answer:**

In [10]:
# 1.Examples of Homogeneous lists include:
list = [1, 2, 3, 8, 33]
print("list of integer: ",list)

list1 = ['dog', 'cat', 'goat']
print("list of animal: ",list1)

list2 = ['John', 'Travis', 'Sheila']
print("list of name: ",list2)

list3 = [2.2, 4.5, 9.8, 10.4]
print("list of floating numbers: ",list3)

list of integer:  [1, 2, 3, 8, 33]
list of animal:  ['dog', 'cat', 'goat']
list of name:  ['John', 'Travis', 'Sheila']
list of floating numbers:  [2.2, 4.5, 9.8, 10.4]


In [12]:
# 2.Examples of Heterogeneous lists include:
list1 = [2, 'cat', 34.33, 'Travis']
print("List of Heterogeneous: ",list1)

list2 = [2.22, 33, 'pen']
print("List of Heterogeneous: ",list2)

List of Heterogeneous:  [2, 'cat', 34.33, 'Travis']
List of Heterogeneous:  [2.22, 33, 'pen']


In [13]:
# 3.Examples of Homogeneous tuple include:
# Different types of tuples

# Empty tuple
my_tuple = ()
print(my_tuple)

# Tuple having integers
# Homogeneous tuple
my_tuple = (1, 2, 3)
print(my_tuple)

# tuple with mixed datatypes
my_tuple = (1, "Hello", 3.4)
print(my_tuple)

# nested tuple
my_tuple = ("mouse", [8, 4, 6], (1, 2, 3))
print(my_tuple)

()
(1, 2, 3)
(1, 'Hello', 3.4)
('mouse', [8, 4, 6], (1, 2, 3))


**Q.8. Explain the mutable and immutable data types with proper explanation & examples.**

**Answer:**

|Mutable|Immutable|
|-------|---------|
|1.A mutable object can be changed after it is created. | 1.An immutable object cannot be changed after it is created.|
|2.Examples : List, Set, Dictionary.| 2.Exg: tuples, int, float, bool, frozenset. |
|3.Mutable objects are not considered as thread-safe in nature. |3.Immutable objects are regarded as thread-safe innature.|
|4.Mutable Objects are slower to access, as compared to immutable objects. | 4.Immutable objects are faster to access when compared to mutable objects. |
|5.Mutable objects are useful when we need to change the size or contents of our object. | 5.Immutable objects are best suitable when we are sure that we don't need to change them at any point in time. |<br>


**Mutable Data Types in Python:**<br>

There are basically 3 mutable data types in Python:

**1.List**<br>
**2.Dictionary**<br>
**3.Set**<br>

**1.List:** A list data structure is a ordered sequence of elements in Python, that is mutable, or changeable. This is because we can change the value of a list in Python by assigning a new value to our already existing list. We need not create any separate (or new) copy of the list to perform these operations. A few examples for the same are given below.

In [14]:
# example to demonstrate list is a mutable data type in python

# our current list
my_list = [1,2,3,4,5]

# using append operation in our list
my_list.append(10)
# printing our list after the operation
print("List after appending a value = ",my_list)

# using extend operation in our list
my_list.extend([6,11,23])
# printing our list after the operation
print("List after extending a list = ",my_list)

# after removing a value from our list
my_list.remove(3)
# printing our list after the operation
print("List after removing a value = ",my_list)

List after appending a value =  [1, 2, 3, 4, 5, 10]
List after extending a list =  [1, 2, 3, 4, 5, 10, 6, 11, 23]
List after removing a value =  [1, 2, 4, 5, 10, 6, 11, 23]


**Explanation:** In the above list, we have performed multiple operations to demonstrate that we can change the value of an already existing list

1.Initially we created a dummy list.<br>
2.Then appended a value to our list, and printed the updated list where our added value was reflected.<br>
3.We also performed operations like extending values and removing values from our original list.<br>
4.As an output, all the performed changes were reflected.<br>
5.Hence, we may conclude that lists in Python are mutable.<br>



**2.Dictionary:** Dictionary in Python is an unordered collection of items. Each item of a dictionary has a key/value pair, using which we can access a particular key or value of a dictionary. Keys in dictionaries are unique in nature.<br>

Dictionary is a mutable data type in Python. That means, we can modify, remove or add values of existing items in a dictionary using the assignment operator.<br>

The addition of items can be performed by using the keys of the dictionary. If the item is already present, then simply the value gets updated. Removing elements in a dictionary can be done simply using pop(), popitem(), clear() or del() in Python.<br>

In [15]:
# example to demonstrate dictionary is a mutable data type in python

# our current dictionary
my_dict = {"state":"WB", "Capital":"Kolkata"}

# adding new key-value pair to our dictionary
my_dict['Country'] = "India"
# printing our dictionary after the operation
print("Dictionary after adding a new key-value pair = ",my_dict)

# updating key-value pair in our dictionary
my_dict['state'] = "West Bengal"
# printing our dictionary after the operation
print("Dictionary after updating an existing key value pair = ",my_dict)

# removing key-value pair in our dictionary
my_dict.pop('Capital')
# printing our dictionary after the operation
print("Dictionary after popping out a key value pair = ",my_dict)

# removing key-value pair in our dictionary
my_dict.clear()
print("After clearing the whole dictionary = ",my_dict)

Dictionary after adding a new key-value pair =  {'state': 'WB', 'Capital': 'Kolkata', 'Country': 'India'}
Dictionary after updating an existing key value pair =  {'state': 'West Bengal', 'Capital': 'Kolkata', 'Country': 'India'}
Dictionary after popping out a key value pair =  {'state': 'West Bengal', 'Country': 'India'}
After clearing the whole dictionary =  {}


**Explanation:** In the above example, we performed operations like adding, updating, and removing elements from our dictionary. This proves that dictionary is a mutable data type in Python.


**3.Set:**
A set is an unordered collection of items in Python. All the elements of the set are unique in nature, that is, there are no duplicates in a set. Also, the elements of the set are immutable in nature, that is, they cannot be changed.However, a set itself is mutable in Python. That means, we can add or remove elements from a set in Python. Hence, we can perform operations in a set that can modify the overall set.<br>

A few notable points regarding the mutability of sets in Python are:<br>
1.Since sets are unordered in nature, we cannot apply indexing to them.<br>
2.We cannot use indexing or slicing to fetch (access) or change the elements of a set. In fact, the set data type itself does not support these operations.<br>
3.However, we may add, update or remove any element from the set. The addition of elements can be performed using the add() method in Python. And, updating operations can be performed using the update() method. We can use the remove() in Python to remove elements from a set.

In [16]:
# example to demonstrate 
# set is a mutable data type in python

# our current set
my_set = {1,2,6,5,7,11}

# adding an element in our set
my_set.add(16)
# printing our set after the operation
print("Set after adding a value : ",my_set)

# adding multiple elements in our set
# multiple elements (such as a list) can be added using update
my_set.update([9,78,100])
# printing our set after the operation
print("Set after updating some values : ",my_set)

# removing element from our set
my_set.remove(2)
# printing our set after the operation
print("Set after removing a value : ",my_set)

Set after adding a value :  {1, 2, 5, 6, 7, 11, 16}
Set after updating some values :  {1, 2, 100, 5, 6, 7, 9, 11, 78, 16}
Set after removing a value :  {1, 100, 5, 6, 7, 9, 11, 78, 16}


**Immutable Data Types in Python:**

There are basically 3 Immutable data types in Python:

**1.Numbers.**<br>
**2.Strings.**<br>
**3.Tuples.**<br>

**1.Numbers:** Numbers are the first and most common example of immutable built-in types in Python. You’ll have no way to access the interior of a numeric value. Because of this limitation, you won’t be able to mutate its value without altering its identity. This is true for all the built-in numeric types—int, float, and complex—as you already learned.

In Python, when you want to use a given object throughout your code, you typically assign that object to a variable and then reuse that variable whenever you need to. Now consider the following example of a variable that points to a numeric value:

In [19]:
number = 314
number

314

In [20]:
id(number)

1716174281520

In above example, you use an assignment statement to create a new variable that holds a reference to an integer number. The id() function returns the number’s identity.

Now say that you don’t know how Python variables work and you think I can probably change the value of this integer variable using another assignment:

In [21]:
number = 7
number

7

In [22]:
id(number)

140703707375584

If you compare the return value of id() in this example with the value you got in the previous example, then you’ll note that they’re different. So, instead of updating the value of number, you’ve created a completely new value and made your variable refer to it.<br>

Once you’ve changed number to refer to 7, then the old 314 gets garbage-collected, which frees up the memory assigned to storing the object.<br>

When it comes to numbers, Python just allows you to reassign the variable name to point to a new numeric value that lives at a completely different memory address. It doesn’t allow you to change the value stored at the original memory address. That’s why 314 has the identity 1716174281520, and 7 has the identity 140703707375584. This behavior is the base of immutability in numeric types.<br>

**2.Strings:**
Python strings are sequences of individual characters. Because they’re sequences, you can efficiently access any individual character in a given string using the indexing operator ([]) and a zero-based integer index:

In [25]:
greeting = "Hello!"

In [26]:
greeting[0]

'H'

In [27]:
greeting[1]

'e'

In [28]:
greeting[2]

'l'

In these examples, you use indexing on a string object to access individual characters. Because you can access characters this way, you can also use id() to learn the identity of each character:

In [29]:
id(greeting[0])

1716104964976

In [30]:
id(greeting[1])

1716096247408

In [31]:
id(greeting[2])

1716096351984

As you can see in these examples, every unique character in a string has its own identity, which is completely independent of the identity of the string object itself:

In [32]:
id(greeting)

1716175420400

With this result, it may seem like Python strings can be mutated in place. However, they can’t. Strings are immutable objects in Python.

**3.Tuples:**
Python’s tuples are another example of an immutable collection of data. You can’t modify them once created. Like strings, tuples are sequences. However, unlike strings, tuples allow you to store any type of object rather than just characters. To create a tuple object, you can use a literal that consists of a comma-separated series of values:

In [33]:
letters = ("a", "b", "c", "d")
type(letters)

tuple

In this example, you create a letters tuple using a literal. This specific tuple holds string objects. Note how the type() function says that your letters variable refers to an instance of the tuple class.

Like with strings and other sequences, you can use the indexing operator to access individual items from an existing tuple. However, you can’t use indexing on the left side of an assignment because tuples are immutable:

In [34]:
letters[0]

'a'

In [35]:
letters[1]

'b'

In [36]:
 letters[2]

'c'

In [38]:
letters[0] = "A"

TypeError: 'tuple' object does not support item assignment

In these examples, note that you can access the items of a tuple object using integer indices that start from 0. However, you can’t use this syntax to change the value of a particular item in the underlying tuple. This operation is forbidden and raises a TypeError.

**Q.9. Write a code to create the given structure using only for loop.**

In [46]:
for i in range(1,6):
    for k in range(1,6-i):
        print(" ",end="")
    for j in range(1,(2*i-1)+1):
        print("*",end="")
    print("\n")

    *

   ***

  *****

 *******

*********



In [50]:
n = int(input("enter the number of row: "))
for i in range(n):
    for j in range(i):
        print(" ",end="")
    for j in range(2*(n-i)-1):
        print("|",end="")
    print()

enter the number of row: 5
|||||||||
 |||||||
  |||||
   |||
    |


**Q.10. Write a code to create the given structure using while loop.**

In [ ]:
|||||||||
 |||||||
  |||||
   |||
    |


In [1]:
a = int(input("Enter number of row: "))
row = 0
while row < a:
    space=0
    
    while space<row:
        print(" ",end="")
        space = space + 1
    star = 0
    
    while star < (a-row):
        print("|", end = "")
        star=star+1
        
    row=row+1
    a=a-1
    
    
    print()

Enter number of row: 9
|||||||||
 |||||||
  |||||
   |||
    |


In [ ]:
n= int(input("Enter Number of Row: "))
i = 1
while(n>0):
    b=1
    while(b<i):
        print(" ",end="")
        b=b+1
    j=1
    while(j<=(n*2)-i):
        print("|",end="")
    print()
    n=n-1
    i=i+1
    
#  In above progream get error

In [55]:
def pattern(n):
    i = n
    while i>=1:
        print(" "*(n-i) + "| " * i)
        i-=1

n= int(input("Enter the number of rows: "))

pattern(n)

Enter the number of rows: 5
| | | | | 
 | | | | 
  | | | 
   | | 
    | 
